src/session/content/message_row/document.rs

```diff
@@ -0,0 +1,357 @@
+            Downloading(progress)
+        } else if remote.is_uploading_active {
+            let progress = remote.uploaded_size as f64 / size as f64;
+            Uploading(progress)
+        } else if local.is_downloading_completed {
+            Downloaded
+        } else if local.can_be_downloaded {
+            CanBeDownloaded
+        } else {
+            dbg!(file);
+            unimplemented!("unknown file status");
+        }
+    }
+}
+
+impl MessageDocument {
+    fn update_document(&self, message: &Message) {
+        if let MessageContent::MessageDocument(data) = message.content().0 {
+            let imp = self.imp();
+
+            let message_text = parse_formatted_text(data.caption);
+            imp.content_label.set_visible(!message_text.is_empty());
+            imp.content_label.set_label(message_text);
+
+            imp.file_name_label.set_label(&data.document.file_name);
+
+            let session = message.chat().session();
+
+            self.update_status(data.document.document, session);
+        }
+    }
+
+    fn update_status(&self, file: File, session: Session) -> FileStatus {
+        let status = FileStatus::from(&file);
+
+        let size = file.size.max(file.expected_size) as u64;
+
+        self.update_size_label(&status, size);
+        self.update_button(file, session, &status);
+
+        status
+    }
+
+    fn block_click(&self, block: bool) {
+        let imp = self.imp();
+        let click = &*imp.click;
+        let document_box = &*imp.document_box;
+
+        let handler_id = imp.click_handler_id.get().unwrap();
+        if block {
+            click.block_signal(handler_id);
+            document_box.remove_css_class("activatable");
+        } else {
+            click.unblock_signal(handler_id);
+            document_box.add_css_class("activatable");
+        }
+    }
+
+    fn update_button(&self, file: File, session: Session, status: &FileStatus) {
+        let imp = self.imp();
+        let button = &*imp.file_status_button;
+        let file_id = file.id;
+
+        let handler_id = match *status {
+            Downloading(_progress) | Uploading(_progress) => {
+                return;
+                // Show loading indicator on button
+            }
+            CanBeDownloaded => {
+                // Download file
+                self.block_click(false);
+                button.set_icon_name("document-save-symbolic");
+                button.connect_clicked(clone!(@weak self as obj, @weak session => move |button| {
+                    let (sender, receiver) = glib::MainContext::sync_channel::<File>(Default::default(), 5);
+                    receiver.attach(
+                        None,
+                        clone!(@weak obj, @weak session => @default-return glib::Continue(false), move |file| {
+                            glib::Continue(obj.update_status(file, session) != FileStatus::Downloaded)
+                    }));
```

Comment:
I discovered a small problem that can be reproduced by the followiung steps:
1. Telegrand is running
2. You send a file from another client to "Saved Messages"
3. You download the file within Telegrand
4. You switch to another chat and then back to "Saved Messages"
5. The message shows the download icon again despite it has been downloaded in step 3

The problem is that tdlib does not send us the `MessageContent` update when the file is downloaded. We can observe the same thing in `photo.rs`. But as we here download the file automatically, we always land here for those photos that we have received while telegrand is running:

https://github.com/melix99/telegrand/blob/088f46766c24592630308ba67d92a1659a21eb77/src/session/content/message_row/photo.rs#L188

One that works is to update the content when we receive a file download update. The `Message::set_content` has to be made `pub(crate)` for this.

Btw, you don't need to check whether the file is downloaded for `glib::Continue`. The channel disappears by itself. So you can also remove the return type of `update_status`

@melix99 any other idea?

```suggestion
                button.connect_clicked(clone!(@weak self as obj, @weak session => move |button| {
                    let (sender, receiver) = glib::MainContext::sync_channel::<File>(Default::default(), 5);
                    receiver.attach(
                        None,
                        clone!(@weak obj => @default-return glib::Continue(false), move |file| {
                            let message = obj.message();
                            if let MessageContent::MessageDocument(mut data) = message.content().0 {
                                data.document.document = file;
                                message.set_content(BoxedMessageContent(MessageContent::MessageDocument(data)));
                            }
                            glib::Continue(true)
                    }));
```

Reply:
Yeah, I think that TDLib expects us to override the file when we download it. So, I plan to create an actual "File" GObject and add a download function to it and we can react to the property changes directly instead of using a channel, this way it will also automatically update the data automatically. Then we also need to create manual GObject for each message content type to allow to have this custom File GObject as a property. Actually I already have an initial implementation of this somewhere in my home pc, so we can do that later to not block this PR.